pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -103,41 +103,56 @@ func (endpointSliceStrategy) AllowUnconditionalUpdate() bool {
 	return true
 }
 
-// dropDisabledConditionsOnCreate will drop the terminating condition if the
-// EndpointSliceTerminatingCondition is disabled. Otherwise the field is left untouched.
-func dropDisabledConditionsOnCreate(endpointSlice *discovery.EndpointSlice) {
-	if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) {
-		return
-	}
-
-	// Always drop the serving/terminating conditions on create when feature gate is disabled.
-	for i := range endpointSlice.Endpoints {
-		endpointSlice.Endpoints[i].Conditions.Serving = nil
-		endpointSlice.Endpoints[i].Conditions.Terminating = nil
+// dropDisabledConditionsOnCreate will drop any fields that are disabled.
+func dropDisabledFieldsOnCreate(endpointSlice *discovery.EndpointSlice) {
+	dropNodeName := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
+	dropTerminating := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition)
+
+	if dropNodeName || dropTerminating {
+		for i := range endpointSlice.Endpoints {
+			if dropNodeName {
+				endpointSlice.Endpoints[i].NodeName = nil
+			}
+			if dropTerminating {
+				endpointSlice.Endpoints[i].Conditions.Serving = nil
+				endpointSlice.Endpoints[i].Conditions.Terminating = nil
+			}
+		}
 	}
 }
 
-// dropDisabledConditionsOnUpdate will drop the terminating condition field if the EndpointSliceTerminatingCondition
-// feature gate is disabled unless an existing EndpointSlice object has the field already set. This ensures
-// the field is not dropped on rollback.
-func dropDisabledConditionsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
-	if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) {
-		return
+// dropDisabledFieldsOnUpdate will drop any disable fields that have not already
+// been set on the EndpointSlice.
+func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
+	dropNodeName := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
```

Comment:
I think this will drop existing resources with the field already set when the feature gate is off, which I think is not desirable.

Reply:
nvm ignore me, I misread :)

Reply:
I may not be understanding this comment correctly, but right below this line I set this to false if `NodeName` has been set on any endpoint:

```
	if dropNodeName {
		for _, ep := range oldEPS.Endpoints {
			if ep.NodeName != nil {
				dropNodeName = false
				break
			}
		}
	}
```

Reply:
I was about to link to the test that covered this and then I realized the gate was accidentally on in the test. That should be updated now and the corrected test does still pass.